Getting Unknown authType: GENERIC when connecting to postgres

### Bug Description

When trying to connect to postgresql running on cloudSQL I am getting the following error:

`javax.net.ssl.SSLHandshakeException: Unknown authType: GENERIC`

I came across this issue when looking at the stacktrace: https://github.com/google/conscrypt/issues/1033 but I couldnt find any workaround.

The code is running perfectly when executed on my Mac, but when running on DataProc I am getting this error.

### Example code (or command)

_No response_

### Stacktrace

```bash
2024-05-15 17:00:30,263 DEBUG hikari.HikariConfig: HikariPool-1 - configuration:
2024-05-15 17:00:30,267 DEBUG hikari.HikariConfig: allowPoolSuspension.............false
2024-05-15 17:00:30,267 DEBUG hikari.HikariConfig: autoCommit......................true
2024-05-15 17:00:30,267 DEBUG hikari.HikariConfig: catalog.........................none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: connectionInitSql...............none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: connectionTestQuery.............none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: connectionTimeout...............30000
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: dataSource......................none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: dataSourceClassName.............none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: dataSourceJNDI..................none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: dataSourceProperties............{password=<masked>, cloudSqlInstance=CloudSqlConnectionString, socketFactory=com.google.cloud.sql.postgres.SocketFactory}
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: driverClassName.................none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: healthCheckProperties...........{}
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: healthCheckRegistry.............none
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: idleTimeout.....................600000
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: initializationFailFast..........true
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: initializationFailTimeout.......1
2024-05-15 17:00:30,268 DEBUG hikari.HikariConfig: isolateInternalQueries..........false
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: jdbc4ConnectionTest.............false
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: jdbcUrl........................."jdbc:postgresql://someip:5432/db"
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: leakDetectionThreshold..........0
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: maxLifetime.....................1800000
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: maximumPoolSize.................1
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: metricRegistry..................none
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: metricsTrackerFactory...........none
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: minimumIdle.....................1
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: password........................<masked>
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: poolName........................"HikariPool-1"
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: readOnly........................false
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: registerMbeans..................false
2024-05-15 17:00:30,269 DEBUG hikari.HikariConfig: scheduledExecutor...............none
2024-05-15 17:00:30,270 DEBUG hikari.HikariConfig: scheduledExecutorService........internal
2024-05-15 17:00:30,270 DEBUG hikari.HikariConfig: threadFactory...................internal
2024-05-15 17:00:30,270 DEBUG hikari.HikariConfig: transactionIsolation............default
2024-05-15 17:00:30,270 DEBUG hikari.HikariConfig: username........................"masked"
2024-05-15 17:00:30,270 DEBUG hikari.HikariConfig: validationTimeout...............5000
2024-05-15 17:00:30,272 INFO hikari.HikariDataSource: HikariPool-1 - Starting...
2024-05-15 17:00:30,388 DEBUG core.InternalConnectorRegistry: First Cloud SQL connection, generating RSA key pair.
2024-05-15 17:00:30,460 DEBUG core.Connector: [CloudSqlConnectionString] Connection info added to cache.
2024-05-15 17:00:30,467 DEBUG core.Refresher: [CloudSqlConnectionString] Force Refresh: the next refresh operation was cancelled. Scheduling new refresh operation immediately.
2024-05-15 17:00:30,467 DEBUG core.Refresher: [CloudSqlConnectionString] Refresh Operation: Acquiring rate limiter permit.
2024-05-15 17:00:30,478 DEBUG core.Refresher: [CloudSqlConnectionString] Refresh Operation: Rate limiter permit acquired.
2024-05-15 17:00:30,687 DEBUG core.DefaultConnectionInfoRepository: [CloudSqlConnectionString] METADATA DONE
2024-05-15 17:00:30,712 DEBUG core.DefaultConnectionInfoRepository: [CloudSqlConnectionString 18] CERT DONE
2024-05-15 17:00:30,879 DEBUG core.DefaultConnectionInfoRepository: [CloudSqlConnectionString 21] SSL CONTEXT
2024-05-15 17:00:30,879 DEBUG core.DefaultConnectionInfoRepository: [CloudSqlConnectionString] INSTANCE DATA DONE
2024-05-15 17:00:30,880 DEBUG core.DefaultConnectionInfoRepository: [CloudSqlConnectionString] ALL FUTURES DONE
2024-05-15 17:00:30,881 DEBUG core.Refresher: [CloudSqlConnectionString] Refresh Operation: Completed refresh with new certificate expiration at 2024-05-15T18:00:30Z.
2024-05-15 17:00:30,882 DEBUG core.Refresher: [CloudSqlConnectionString] Refresh Operation: Next operation scheduled at 2024-05-15T17:56:29Z.
2024-05-15 17:00:30,887 DEBUG core.Refresher: [CloudSqlConnectionString] Now = 2024-05-15T17:00:30.886740Z, Current client certificate expiration = 2024-05-15T18:00:30Z
2024-05-15 17:00:30,887 DEBUG core.Connector: [192.168.1.1] Connecting to instance.
2024-05-15 17:00:30,901 DEBUG core.Connector: TLS handshake failed!
2024-05-15 17:00:30,902 DEBUG core.Connector: [CloudSqlConnectionString] Socket connection failed! Trigger a refresh.
2024-05-15 17:00:30,902 DEBUG core.Refresher: [CloudSqlConnectionString] Force Refresh: the next refresh operation was cancelled. Scheduling new refresh operation immediately.
2024-05-15 17:00:30,902 DEBUG core.Refresher: [CloudSqlConnectionString] Refresh Operation: Acquiring rate limiter permit.
2024-05-15 17:00:30,909 DEBUG pool.PoolBase: HikariPool-1 - Failed to create/setup connection: The connection attempt failed.
2024-05-15 17:00:30,910 DEBUG pool.HikariPool: HikariPool-1 - Cannot acquire connection from data source
org.postgresql.util.PSQLException: The connection attempt failed.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:358)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:54)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:273)
	at org.postgresql.Driver.makeConnection(Driver.java:446)
	at org.postgresql.Driver.connect(Driver.java:298)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:112)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:118)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:341)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:193)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:428)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:499)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:112)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:72)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.google.cloud.hadoop.services.agent.job.shim.HadoopRunClassShim.main(HadoopRunClassShim.java:19)
Caused by: javax.net.ssl.SSLHandshakeException: Unknown authType: GENERIC
	at org.conscrypt.SSLUtils.toSSLHandshakeException(SSLUtils.java:361)
	at org.conscrypt.ConscryptEngine.convertException(ConscryptEngine.java:1138)
	at org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1093)
	at org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:880)
	at org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:751)
	at org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:716)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:833)
	at org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(ConscryptEngineSocket.java:706)
	at org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:230)
	at org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:209)
	at com.google.cloud.sql.core.Connector.connect(Connector.java:122)
	at com.google.cloud.sql.core.InternalConnectorRegistry.connect(InternalConnectorRegistry.java:179)
	at com.google.cloud.sql.postgres.SocketFactory.createSocket(SocketFactory.java:81)
	at org.postgresql.core.PGStream.createSocket(PGStream.java:231)
	at org.postgresql.core.PGStream.<init>(PGStream.java:98)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:136)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:262)
	... 30 more
Caused by: java.security.cert.CertificateException: Unknown authType: GENERIC
	at java.base/sun.security.validator.EndEntityChecker.checkTLSServer(EndEntityChecker.java:289)
	at java.base/sun.security.validator.EndEntityChecker.check(EndEntityChecker.java:144)
	at java.base/sun.security.validator.Validator.validate(Validator.java:277)
	at java.base/sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:222)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:129)
	at org.conscrypt.ConscryptEngineSocket$2.checkServerTrusted(ConscryptEngineSocket.java:156)
	at org.conscrypt.Platform.checkServerTrusted(Platform.java:330)
	at org.conscrypt.ConscryptEngine.verifyCertificateChain(ConscryptEngine.java:1643)
	at org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
	at org.conscrypt.NativeSsl.readDirectByteBuffer(NativeSsl.java:567)
	at org.conscrypt.ConscryptEngine.readPlaintextDataDirect(ConscryptEngine.java:1099)
	at org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1083)
	... 44 more
```


### Steps to reproduce?

1. ?
2. ?
3. ?
...


### Environment

1. OS type and version: DataProc running image 2.2.16-ubuntu22
2. Java SDK version:  11
3. Cloud SQL Java Socket Factory version: latest


### Additional Details

_No response_

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -229,5 +229,23 @@
       </plugin>
     </plugins>
   </build>
+  <profiles>
+    <profile>
+      <id>google-conscript</id>
+      <dependencies>
+        <dependency>
+          <groupId>org.conscrypt</groupId>
+          <artifactId>conscrypt-openjdk-uber</artifactId>
+          <version>2.5.2</version>
+        </dependency>
+        <dependency>
+          <groupId>org.bouncycastle</groupId>
+          <artifactId>bcpkix-jdk15on</artifactId>
+          <version>1.70</version>
+          <scope>provided</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+  </profiles>
 
 </project>

--- a/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundDelegatingTrustManger.java
+++ b/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundDelegatingTrustManger.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import java.net.Socket;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+/**
+ * This is a workaround for a known bug in Conscrypt crypto in how it handles X509 auth type.
+ * OpenJDK interpres the X509 certificate auth type as "UNKNOWN" while Conscrypt interpret the same
+ * certificate as auth type "GENERIC". This incompatibility causes problems in the JDK.
+ *
+ * <p>This adapter works around the issue by creating wrappers around all TrustManager instances. It
+ * replaces "GENERIC" auth type with "UNKNOWN" auth type before delegating calls.
+ *
+ * <p>See https://github.com/google/conscrypt/issues/1033#issuecomment-982701272
+ */
+class ConscryptWorkaroundDelegatingTrustManger extends X509ExtendedTrustManager {
+  private final X509ExtendedTrustManager tm;
+
+  ConscryptWorkaroundDelegatingTrustManger(X509ExtendedTrustManager tm) {
+    this.tm = tm;
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    tm.checkClientTrusted(chain, fixAuthType(authType), socket);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    tm.checkClientTrusted(chain, fixAuthType(authType), engine);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    tm.checkClientTrusted(chain, fixAuthType(authType));
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    tm.checkServerTrusted(chain, fixAuthType(authType), socket);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    tm.checkServerTrusted(chain, fixAuthType(authType), engine);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    tm.checkServerTrusted(chain, fixAuthType(authType));
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return tm.getAcceptedIssuers();
+  }
+
+  private String fixAuthType(String authType) {
+    if ("GENERIC".equals(authType)) {
+      return "UNKNOWN";
+    }
+    return authType;
+  }
+}

--- a/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundTrustManagerFactory.java
+++ b/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundTrustManagerFactory.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.Security;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * This is a workaround for a known bug in Conscrypt crypto in how it handles X509 auth type.
+ * OpenJDK interpres the X509 certificate auth type as "UNKNOWN" while Conscrypt interpret the same
+ * certificate as auth type "GENERIC". This incompatibility causes problems in the JDK.
+ *
+ * <p>This adapter works around the issue by creating wrappers around all TrustManager instances. It
+ * replaces "GENERIC" auth type with "UNKNOWN" auth type before delegating calls.
+ *
+ * <p>In the JVM, we need to implement 3 classes to make sure that we are capturing all of the
+ * TrustManager instances created when Conscrypt is the Java Crypto provider so that we can wrap
+ * them with the ConscryptWorkaroundTrustManager:
+ *
+ * <p>class ConscryptWorkaroundTrustManagerFactory extends TrustManagerFactory - has a bunch of
+ * final methods that delegate to a TrustManagerFactorySpi. class
+ * ConscryptWorkaroundTrustManagerFactorySpi implements TrustManagerFactorySpi - can actually
+ * intercept and delegate calls related to trust managers and wrap them with
+ * ConscryptWorkaroundTrustManager ConscryptWorkaroundTrustManager - the workaround for the
+ * Conscrypt bug.
+ *
+ * <p>See https://github.com/google/conscrypt/issues/1033#issuecomment-982701272
+ */
+class ConscryptWorkaroundTrustManagerFactory extends TrustManagerFactory {
+  private static final boolean CONSCRYPT_TLS;
+
+  static {
+    // Provider name is "Conscrypt", hardcoded string in the library source:
+    // https://github.com/google/conscrypt/blob/655ad5069e1cb4d1989b8117eaf090371885af99/openjdk/src/main/java/org/conscrypt/Platform.java#L149
+    Provider p = Security.getProvider("Conscrypt");
+    if (p != null) {
+      try {
+        SSLContext ctx = SSLContext.getInstance("TLS");
+        Provider prov = ctx.getProvider();
+        CONSCRYPT_TLS = "Conscrypt".equals(prov.getName());
+      } catch (NoSuchAlgorithmException e) {
+        throw new RuntimeException("Unable to load algorithm TLS", e);
+      }
+    } else {
+      CONSCRYPT_TLS = false;
+    }
+  }
+
+  /** Returns true if the Conscrypt Java Crypto Extension is installed. */
+  static boolean isWorkaroundNeeded() {
+    return CONSCRYPT_TLS;
+  }
+
+  static ConscryptWorkaroundTrustManagerFactory newInstance() throws NoSuchAlgorithmException {
+    TrustManagerFactory delegate = TrustManagerFactory.getInstance("X.509");
+    return new ConscryptWorkaroundTrustManagerFactory(delegate);
+  }
+
+  private ConscryptWorkaroundTrustManagerFactory(TrustManagerFactory delegate) {
+    super(
+        new ConscryptWorkaroundTrustManagerFactorySpi(delegate),
+        delegate.getProvider(),
+        delegate.getAlgorithm());
+  }
+}

--- a/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundTrustManagerFactorySpi.java
+++ b/core/src/main/java/com/google/cloud/sql/core/ConscryptWorkaroundTrustManagerFactorySpi.java
@@ -1,0 +1,71 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import java.security.InvalidAlgorithmParameterException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import javax.net.ssl.ManagerFactoryParameters;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.TrustManagerFactorySpi;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+/**
+ * This is a workaround for a known bug in Conscrypt crypto in how it handles X509 auth type.
+ * OpenJDK interpres the X509 certificate auth type as "UNKNOWN" while Conscrypt interpret the same
+ * certificate as auth type "GENERIC". This incompatibility causes problems in the JDK.
+ *
+ * <p>This adapter works around the issue by creating wrappers around all TrustManager instances. It
+ * replaces "GENERIC" auth type with "UNKNOWN" auth type before delegating calls.
+ *
+ * <p>See https://github.com/google/conscrypt/issues/1033#issuecomment-982701272
+ */
+class ConscryptWorkaroundTrustManagerFactorySpi extends TrustManagerFactorySpi {
+  private final TrustManagerFactory delegate;
+
+  ConscryptWorkaroundTrustManagerFactorySpi(TrustManagerFactory delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  protected void engineInit(KeyStore ks) throws KeyStoreException {
+    delegate.init(ks);
+  }
+
+  @Override
+  protected void engineInit(ManagerFactoryParameters spec)
+      throws InvalidAlgorithmParameterException {
+    delegate.init(spec);
+  }
+
+  @Override
+  protected TrustManager[] engineGetTrustManagers() {
+    TrustManager[] tms = delegate.getTrustManagers();
+    TrustManager[] delegates = new TrustManager[tms.length];
+    for (int i = 0; i < tms.length; i++) {
+      if (tms[i] instanceof X509ExtendedTrustManager) {
+        delegates[i] =
+            new ConscryptWorkaroundDelegatingTrustManger((X509ExtendedTrustManager) tms[i]);
+      } else {
+
+        delegates[i] = tms[i];
+      }
+    }
+    return delegates;
+  }
+}

--- a/core/src/main/java/com/google/cloud/sql/core/DefaultConnectionInfoRepository.java
+++ b/core/src/main/java/com/google/cloud/sql/core/DefaultConnectionInfoRepository.java
@@ -382,8 +382,20 @@ class DefaultConnectionInfoRepository implements ConnectionInfoRepository {
       KeyStore trustedKeyStore = KeyStore.getInstance(KeyStore.getDefaultType());
       trustedKeyStore.load(null, null);
       trustedKeyStore.setCertificateEntry("instance", instanceMetadata.getInstanceCaCertificate());
-      TrustManagerFactory tmf = TrustManagerFactory.getInstance("X.509");
+      TrustManagerFactory tmf;
+
+      // Note: This is a workaround for Conscrypt bug #1033
+      // Conscrypt is the JCE provider on some Google Cloud runtimes like DataProc.
+      // https://github.com/google/conscrypt/issues/1033
+      //
+      if (ConscryptWorkaroundTrustManagerFactory.isWorkaroundNeeded()) {
+        tmf = ConscryptWorkaroundTrustManagerFactory.newInstance();
+      } else {
+        tmf = TrustManagerFactory.getInstance("X.509");
+      }
+
       tmf.init(trustedKeyStore);
+
       SSLContext sslContext;
 
       try {

--- a/jdbc/postgres/pom.xml
+++ b/jdbc/postgres/pom.xml
@@ -104,7 +104,44 @@
         </dependency>
       </dependencies>
     </profile>
+
+    <profile>
+      <id>google-conscript</id>
+      <dependencies>
+        <dependency>
+          <groupId>org.conscrypt</groupId>
+          <artifactId>conscrypt-openjdk-uber</artifactId>
+          <version>2.5.2</version>
+        </dependency>
+      </dependencies>
+      <build>
+        <plugins>
+          <plugin>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <version>3.2.5</version>
+            <executions>
+              <execution>
+                <phase>integration-test</phase>
+                <goals>
+                  <goal>test</goal>
+                </goals>
+                <configuration>
+                  <!-- double-equals means override the JRE default properties, don't append -->
+                  <argLine>-Djava.security.properties==src/test/resources/conscrypt-security.properties</argLine>
+                  <classpathDependencyExcludes>
+                    <classpathDependencyExclude>org.bouncycastle:bcpkix-jdk15on</classpathDependencyExclude>
+                    <classpathDependencyExclude>org.bouncycastle:bcprov-jdk15on</classpathDependencyExclude>
+                    <classpathDependencyExclude>org.bouncycastle:bcutil-jdk15on</classpathDependencyExclude>
+                  </classpathDependencyExcludes>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
   </profiles>
+
 
   <build>
     <plugins>

--- a/jdbc/postgres/src/test/java/com/google/cloud/sql/postgres/JdbcPostgresIamAuthIntegrationTests.java
+++ b/jdbc/postgres/src/test/java/com/google/cloud/sql/postgres/JdbcPostgresIamAuthIntegrationTests.java
@@ -22,11 +22,15 @@ import static com.google.common.truth.Truth.assertWithMessage;
 import com.google.common.collect.ImmutableList;
 import com.zaxxer.hikari.HikariConfig;
 import com.zaxxer.hikari.HikariDataSource;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.Security;
 import java.sql.*;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Properties;
 import java.util.concurrent.TimeUnit;
+import javax.net.ssl.SSLContext;
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Rule;
@@ -34,9 +38,13 @@ import org.junit.Test;
 import org.junit.rules.Timeout;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 @RunWith(JUnit4.class)
 public class JdbcPostgresIamAuthIntegrationTests {
+  private static final Logger log =
+      LoggerFactory.getLogger(JdbcPostgresIamAuthIntegrationTests.class);
 
   // [START cloud_sql_connector_postgres_jdbc_iam_auth]
   private static final String CONNECTION_NAME = System.getenv("POSTGRES_IAM_CONNECTION_NAME");
@@ -59,6 +67,19 @@ public class JdbcPostgresIamAuthIntegrationTests {
                         "Environment variable '%s' must be set to perform these tests.", varName))
                 .that(System.getenv(varName))
                 .isNotEmpty());
+
+    log.info("Crypto Providers: ");
+    Provider[] providers = Security.getProviders();
+    for (Provider p : providers) {
+      log.info("  {}", p.getName());
+    }
+    try {
+      SSLContext ctx = SSLContext.getInstance("TLS");
+      Provider prov = ctx.getProvider();
+      log.info("TLS Provider: {}", prov.getName());
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
   }
 
   @Before

--- a/jdbc/postgres/src/test/resources/conscrypt-security.properties
+++ b/jdbc/postgres/src/test/resources/conscrypt-security.properties
@@ -1,0 +1,146 @@
+####
+# This file fully replaces the built-in JVM security properties for testing
+# the Conscrypt crypto library.
+
+
+###
+# Updated properties to prioritize the Conscrypt provider over
+# OpenJDK providers.
+
+# Make Conscrypt the #1 security provider
+security.provider.1=org.conscrypt.OpenSSLProvider
+
+# Allow the JVM to fall back to default OpenJDK security providers
+# in case Conscrypt does not include an algorithm.
+security.provider.2=SUN
+security.provider.3=SunRsaSign
+security.provider.4=SunEC
+security.provider.5=SunJSSE
+security.provider.6=SunJCE
+security.provider.7=SunJGSS
+security.provider.8=SunSASL
+security.provider.9=XMLDSig
+security.provider.10=SunPCSC
+security.provider.11=JdkLDAP
+security.provider.12=JdkSASL
+security.provider.13=SunPKCS11
+
+###
+# JVM default property values. These properties were copied from
+# OpenJDK 1.22.3 Linux.
+
+securerandom.source=file:/dev/random
+
+securerandom.strongAlgorithms=NativePRNGBlocking:SUN,DRBG:SUN
+
+securerandom.drbg.config=
+
+login.configuration.provider=sun.security.provider.ConfigFile
+
+
+policy.provider=sun.security.provider.PolicyFile
+
+policy.url.1=file:${java.home}/conf/security/java.policy
+policy.url.2=file:${user.home}/.java.policy
+
+policy.expandProperties=true
+
+policy.allowSystemProperty=true
+
+policy.ignoreIdentityScope=false
+
+keystore.type=pkcs12
+
+keystore.type.compat=true
+
+package.access=sun.misc.,\
+               sun.reflect.
+
+package.definition=sun.misc.,\
+                   sun.reflect.
+
+security.overridePropertiesFile=true
+
+ssl.KeyManagerFactory.algorithm=SunX509
+ssl.TrustManagerFactory.algorithm=PKIX
+
+
+
+networkaddress.cache.negative.ttl=10
+
+
+
+
+
+
+
+krb5.kdc.bad.policy = tryLast
+
+sun.security.krb5.disableReferrals=false
+
+sun.security.krb5.maxReferrals=5
+
+
+jdk.certpath.disabledAlgorithms=MD2, MD5, SHA1 jdkCA & usage TLSServer, \
+    RSA keySize < 1024, DSA keySize < 1024, EC keySize < 224, \
+    SHA1 usage SignedJAR & denyAfter 2019-01-01
+
+
+jdk.security.legacyAlgorithms=SHA1, \
+    RSA keySize < 2048, DSA keySize < 2048, \
+    DES, DESede, MD5, RC2, ARCFOUR
+
+jdk.jar.disabledAlgorithms=MD2, MD5, RSA keySize < 1024, \
+      DSA keySize < 1024, SHA1 denyAfter 2019-01-01
+
+http.auth.digest.disabledAlgorithms = MD5, SHA-1
+
+jdk.tls.disabledAlgorithms=SSLv3, TLSv1, TLSv1.1, DTLSv1.0, RC4, DES, \
+    MD5withRSA, DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL, \
+    ECDH
+
+jdk.tls.legacyAlgorithms=NULL, anon, RC4, DES, 3DES_EDE_CBC
+
+
+jdk.tls.keyLimits=AES/GCM/NoPadding KeyUpdate 2^37, \
+                  ChaCha20-Poly1305 KeyUpdate 2^37
+
+crypto.policy=unlimited
+
+jdk.xml.dsig.secureValidationPolicy=\
+    disallowAlg http://www.w3.org/TR/1999/REC-xslt-19991116,\
+    maxTransforms 5,\
+    maxReferences 30,\
+    disallowReferenceUriSchemes file http https,\
+    minKeySize RSA 1024,\
+    minKeySize DSA 1024,\
+    minKeySize EC 224,\
+    noDuplicateIds,\
+    noRetrievalMethodLoops
+
+
+
+
+
+jceks.key.serialFilter = java.base/java.lang.Enum;java.base/java.security.KeyRep;\
+  java.base/java.security.KeyRep$Type;java.base/javax.crypto.spec.SecretKeySpec;!*
+
+
+
+
+
+
+
+
+
+
+jdk.sasl.disabledMechanisms=
+
+jdk.security.caDistrustPolicies=SYMANTEC_TLS
+
+jdk.io.permissionsUseCanonicalPath=false
+
+
+
+
+jdk.tls.alpnCharset=ISO_8859_1

--- a/pom.xml
+++ b/pom.xml
@@ -304,7 +304,7 @@
         <scope>test</scope>
       </dependency>
       <dependency>
-      <groupId>org.bouncycastle</groupId>
+        <groupId>org.bouncycastle</groupId>
         <artifactId>bcpkix-jdk15on</artifactId>
         <version>1.70</version>
         <scope>test</scope>
@@ -345,6 +345,14 @@
         <filtering>true</filtering>
       </resource>
     </resources>
+
+    <extensions>
+      <extension>
+        <groupId>kr.motd.maven</groupId>
+        <artifactId>os-maven-plugin</artifactId>
+        <version>1.4.1.Final</version>
+      </extension>
+    </extensions>
 
     <plugins>
 
@@ -900,6 +908,40 @@
                 <goals>
                   <goal>single</goal>
                 </goals>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+
+    <profile>
+      <id>google-conscript</id>
+      <dependencies>
+        <dependency>
+          <groupId>org.conscrypt</groupId>
+          <artifactId>conscrypt-openjdk-uber</artifactId>
+          <version>2.5.2</version>
+        </dependency>
+      </dependencies>
+      <build>
+        <plugins>
+          <plugin>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <version>3.2.5</version>
+            <executions>
+              <execution>
+                <phase>integration-test</phase>
+                <goals>
+                  <goal>test</goal>
+                </goals>
+                <configuration>
+                  <classpathDependencyExcludes>
+                    <classpathDependencyExclude>org.bouncycastle:bcpkix-jdk15on</classpathDependencyExclude>
+                    <classpathDependencyExclude>org.bouncycastle:bcprov-jdk15on</classpathDependencyExclude>
+                    <classpathDependencyExclude>org.bouncycastle:bcutil-jdk15on</classpathDependencyExclude>
+                  </classpathDependencyExcludes>
+                </configuration>
               </execution>
             </executions>
           </plugin>

